### PR TITLE
Add currency attributes to ibetStraightBond

### LINKED ISSUE #546: [FEATURE] Supports multiple currency-denominated tokens
## Is your feature request related to a problem? Please describe.
All token values are currently defined as integers and are assumed to be denominated in Japanese Yen. On the other hand, of course, there are also needs to use assets denominated in other currencies.

Against this background, items related to currency will be added with the following changes to the token contract specifications. We would like to reflect this specification change in the functionality of ibet-prime.
https://github.com/BoostryJP/ibet-SmartContract/issues/423

### PR DESCRIPTION
**Pull request by YoshihitoAso**
close #546

### CODE REVIEW

File: app/model/blockchain/token.py [hunk 1]
```diff
@@ -394,10 +394,14 @@ def force_unlock(
...
 class IbetStraightBondContract(IbetSecurityTokenInterface):
     face_value: int
+    face_value_currency: str
     interest_rate: float
     interest_payment_date: List[str]
+    interest_payment_currency: str
     redemption_date: str
     redemption_value: int
+    redemption_value_currency: str
+    base_fx_rate: float
```
Comment (YoshihitoAso) [inline app/model/blockchain/token.py:404]: Added four new items to the Bond contract.
- face_value_currency
- interest_payment_currency
- redemption_value_currency
- base_fx_rate

File: app/model/blockchain/token.py [hunk 2]
```diff
@@ -495,20 +499,39 @@ def get(self):
 
         # Set IbetStraightBondToken attribute
         self.face_value = ContractUtils.call_function(contract, "faceValue", (), 0)
+        self.face_value_currency = ContractUtils.call_function(
+            contract, "faceValueCurrency", (), ""
+        )
```
Comment (YoshihitoAso) [inline app/model/blockchain/token.py:504]: The default value of `face_value_currency` is "".

File: app/model/blockchain/token.py [hunk 3]
```diff
@@ -495,20 +499,39 @@ def get(self):
...
         self.interest_rate = float(
             Decimal(str(ContractUtils.call_function(contract, "interestRate", (), 0)))
             * Decimal("0.0001")
         )
+        self.interest_payment_currency = ContractUtils.call_function(
+            contract, "interestPaymentCurrency", (), ""
+        )
```
Comment (YoshihitoAso) [inline app/model/blockchain/token.py:511]: The default value of `interest_payment_currency` is "".

File: app/model/blockchain/token.py [hunk 4]
```diff
@@ -495,20 +499,39 @@ def get(self):
...
+        try:
+            _raw_base_fx_rate = ContractUtils.call_function(
+                contract, "baseFXRate", (), ""
+            )
+            if _raw_base_fx_rate is not None and _raw_base_fx_rate != "":
+                self.base_fx_rate = float(_raw_base_fx_rate)
+            else:
+                self.base_fx_rate = 0.0
+        except ValueError:
+            self.base_fx_rate = 0.0
```
Comment (YoshihitoAso) [inline app/model/blockchain/token.py:534]: The default value of `base_fx_rate` is 0.0.
Error handling is performed when invalid data that cannot be converted to float is set.

File: app/model/blockchain/token.py [hunk 5]
```diff
@@ -630,6 +693,46 @@ def update(
...
+            except ContractRevertError:
+                raise
+            except TimeExhausted as timeout_error:
+                raise SendTransactionError(timeout_error)
+            except Exception as err:
+                raise SendTransactionError(err)
+
+        if data.base_fx_rate is not None:
+            tx = contract.functions.setBaseFXRate(
+                str(data.base_fx_rate)
```
Comment (YoshihitoAso) [inline app/model/blockchain/token.py:718]: Convert the input value to str and set it.

File: app/model/schema/ledger.py [hunk 6]
```diff
@@ -132,6 +132,7 @@ class RetrieveLedgerHistoryResponse(BaseModel):
 
     created: str
     token_name: str
+    currency: str
```
Comment (YoshihitoAso) [inline app/model/schema/ledger.py:135]: The default value is "".
Although the item itself does not exist in the data created with the old version, the default value (i.e. "") is returned as the API return value.

File: app/model/schema/token.py [hunk 7]
```diff
@@ -60,6 +67,18 @@ class IbetStraightBondCreate(BaseModel):
...
+    @classmethod
+    def base_fx_rate_6_decimal_places(cls, v):
+        if v is not None:
+            float_data = float(v * 10**6)
+            int_data = int(v * 10**6)
+            if not math.isclose(int_data, float_data):
+                raise ValueError(
+                    "base_fx_rate must be less than or equal to six decimal places"
+                )
+        return v
```
Comment (YoshihitoAso) [inline app/model/schema/token.py:80]: The number of digits after the decimal point is 6.

File: migrations/versions/53caec5f4dce_v23_12_0_feature_546.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+# revision identifiers, used by Alembic.
+revision = "53caec5f4dce"
+down_revision = "8ed94523e21c"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # Delete all `token_cache` data
+    op.get_bind().execute(delete(TokenCache))
```
Comment (YoshihitoAso) [inline migrations/versions/53caec5f4dce_v23_12_0_feature_546.py:22]: If valid token cache data exists, the old format data will be returned, so I decided to delete it all at once.

**Review by purplesmoke05 — APPROVED** [review]
LGTM

**Pull request merged**